### DISCUSSION TITLE
Auth0 role in session

### DISCUSSION THREAD

**Discussion by laurent512** [nextauthjs/next-auth#2551] [Help]
Hello,

I am willing to deploy a role based dashboard ( more options  for _admin_  than _user_) 

The only login provider I have enabled is authO and I have defined roles for my users there.

Is there a way to read it via session.user for example?

Thanks in advance

**Comment by balazsorban44** [comment]
Absolutely. Check out the `jwt` and `session` callbacks: https://next-auth.js.org/configuration/callbacks

**Reply by laurent512** [reply]
Thank you for such a quick response, I highly appreciate !
The concepts of jwt and authentication mechasnisms are new to me.

May I ask for a bit more expalanation? thank you :)

**Reply by balazsorban44** [reply]
Have you read:
https://next-auth.js.org/configuration/callbacks#jwt-callback
https://next-auth.js.org/configuration/callbacks#session-callback

Please tell me if even after reading it wasn't clear.

**Reply by laurent512** [reply]
Thank you for your answer again !
Indeed, I went through that documentation but I do not understand how to update my code according that.
Would you be able to provide some more guidance please ?

**Reply by balazsorban44** [reply]
The JWT callback is used to specify what is available in the session cookie. The `session` callback is used to define what is returned by the `useSession` hook amongst other methods.

Assuming that Auth0 returns a `role` claim, you could do:

```js
...
callbacks: {
  jwt(token, user, account, profile) {
    if(profile.role) token.role = profile.role
    return token
  },
  session(session, token) {
    session.user.role = token.role
    return session
  }
}
```

**Reply by laurent512** [reply]
Thank you so much ! So this would be the resulting code , am I right ? 
Because I have the following error :/ https://ibb.co/tZ7QTcp



```js
import NextAuth from "next-auth";
import Providers from "next-auth/providers";

export default NextAuth({
  providers: [
    Providers.Auth0({
      clientId: process.env.AUTH0_CLIENT_ID,
      clientSecret: process.env.AUTH0_CLIENT_SECRET,
      domain: process.env.AUTH0_DOMAIN,
    }),
  ],
  callbacks: {
    jwt(token, user, account, profile) {
      if (profile.role) token.role = profile.role;
      return token;
    },
    session(session, token) {
      session.user.role = token.role;
      return session;
    },
  },
});
```

**Reply by balazsorban44** [reply]
Sorry, 
```diff
- if (profile.role) token.role = profile.role;
+ if (profile) {
+   token.role = profile.role;
+ }
```

As the above documentation suggests, anything else than `token` will be available only on the first request, that is why you have to persist it on `token`

Also, I don't know for sure that Auth0 gives you a claim called `role`, so you have to check.

**Reply by laurent512** [reply]
Thank you so much for your help, it works now !!
You are correct, I will check on auth0 side how to give the role information ( it is not there by default)
